File: handconsole_adapter/apps/desktop/src/types/frame.ts
export interface Point2D {
  x: number;
  y: number;
}

export type AirInkSourceType = "Camera" | "Mock" | "Playback";

export type AirInkGestureState =
  | "HOVER"
  | "ARMING_DOWN"
  | "INKING"
  | "ARMING_UP"
  | "HAND_LOST";

export interface AirInkFrame {
  timestamp_ms: number;
  frame_id: number;
  source_type: AirInkSourceType;
  camera: {
    width: number;
    height: number;
    fps: number;
  };
  tracking: {
    hand_detected: boolean;
    confidence: number;
    landmarks: Point2D[];
    raw_tip: Point2D | null;
    stable_tip: Point2D | null;
    pinch_ratio: number | null;
    extension_ratio: number | null;
  };
  gesture: {
    state: AirInkGestureState;
    drawing_active: boolean;
  };
  quality: {
    tracking_ok: boolean;
    jump_rejected: boolean;
    frame_dropped: boolean;
  };
}

File: handconsole_adapter/apps/desktop/src-tauri/src/frame.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Point2D {
    pub x: f32,
    pub y: f32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AirInkCameraStatus {
    pub status: String,
    pub camera_index: u32,
    pub width: u32,
    pub height: u32,
    pub fps: f32,
    pub error_message: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AirInkCameraInfo {
    pub width: u32,
    pub height: u32,
    pub fps: f32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AirInkTracking {
    pub hand_detected: bool,
    pub confidence: f32,
    pub landmarks: Vec<Point2D>,
    pub raw_tip: Option<Point2D>,
    pub stable_tip: Option<Point2D>,
    pub pinch_ratio: Option<f32>,
    pub extension_ratio: Option<f32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AirInkGesture {
    pub state: String,
    pub drawing_active: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AirInkQuality {
    pub tracking_ok: bool,
    pub jump_rejected: bool,
    pub frame_dropped: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AirInkFrame {
    pub timestamp_ms: u64,
    pub frame_id: u64,
    pub source_type: String,
    pub camera: AirInkCameraInfo,
    pub tracking: AirInkTracking,
    pub gesture: AirInkGesture,
    pub quality: AirInkQuality,
}
